test/jdk/java/lang/StableValue/StableListTest.java

```diff
@@ -0,0 +1,425 @@
+        AtomicInteger cnt = new AtomicInteger();
+        iterator.forEachRemaining(_ -> cnt.incrementAndGet());
+        assertEquals(0, cnt.get());
+    }
+
+    @Test
+    void iteratorPartial() {
+        var iterator = newList().iterator();
+        for (int i = 0; i < INDEX; i++) {
+            assertTrue(iterator.hasNext());
+            assertTrue(iterator.hasNext());
+            assertEquals(i, iterator.next());
+        }
+        assertTrue(iterator.hasNext());
+        AtomicInteger cnt = new AtomicInteger();
+        iterator.forEachRemaining(_ -> cnt.incrementAndGet());
+        assertEquals(SIZE - INDEX, cnt.get());
+        assertFalse(iterator.hasNext());
+        assertThrows(NoSuchElementException.class, iterator::next);
+    }
+
+    @Test
+    void subList() {
+        var lazy = newList();
+        var lazySubList = lazy.subList(1, SIZE);
+        assertInstanceOf(RandomAccess.class, lazySubList);
+        var regularList = newRegularList();
+        var regularSubList = regularList.subList(1, SIZE);
+        assertEquals(regularSubList, lazySubList);
+    }
+
+    @Test
+    void subList2() {
+        var lazy = newList();
+        var lazySubList = lazy.subList(1, SIZE);
+        lazySubList.get(0);
+        var eq = newList();
+        eq.get(1);
+        assertEquals(eq.toString(), lazy.toString());
+    }
+
+    @Test
+    void subListToString() {
+        subListToString0(newList());
+        subListToString0(newList().subList(1, SIZE));
+        subListToString0(newList().subList(1, SIZE).subList(0, SIZE - 2));
+    }
+
+    void subListToString0(List<Integer> subList) {
+        assertEquals(asString(".unset", subList), subList.toString());
+
+        var first = subList.getFirst();
+        assertEquals(asString(first.toString(), subList), subList.toString());
+    }
+
+    @Test
+    void reversed() {
+        var reversed = newList().reversed();
+        assertInstanceOf(RandomAccess.class, reversed);
+        assertEquals(SIZE - 1, reversed.getFirst());
+        assertEquals(0, reversed.getLast());
+
+        var reversed2 = reversed.reversed();
+        assertEquals(0, reversed2.getFirst());
+        assertEquals(SIZE - 1, reversed2.getLast());
+    }
+
+    @Test
+    void reversedToString() {
+        var reversed = newList().reversed();
+        subListToString0(reversed);
+    }
+
+    @Test
+    void subListReversedToString() {
+        var list = newList().subList(1, SIZE - 1).reversed();
+        // This combination is not lazy. There has to be a limit somewhere.
+        var regularList = newRegularList().subList(1, SIZE - 1).reversed();
+        assertEquals(regularList.toString(), list.toString());
+    }
+
+    @Test
+    void recursiveCall() {
+        AtomicReference<IntFunction<Integer>> ref = new AtomicReference<>();
+        var lazy = StableValue.list(SIZE, i -> ref.get().apply(i));
+        ref.set(lazy::get);
+        assertThrows(IllegalStateException.class, () -> lazy.get(INDEX));
```

Comment:
Perhaps test the exception message to verify that it is indeed a recursion detection?